db-service/lib/SQLService.js

```diff
@@ -201,12 +201,20 @@ class SQLService extends DatabaseService {
     async function deep_delete(/** @type {Request} */ req) {
       const transitions = getTransition(req.query.target, this)
       if (transitions.target !== transitions.queryTarget) {
-        const elements = transitions.queryTarget.keys
-          ? Object.keys(transitions.queryTarget.keys).filter(key => !transitions.queryTarget.keys[key].virtual)
-          : Object.keys(transitions.queryTarget.elements).filter(
-              key => !transitions.queryTarget.elements[key].isAssociation,
-            )
-        const matchedKeys = elements.filter(key => transitions.mapping.has(key)).map(k => ({ ref: [k] }))
+        const keys = []
+        const transitionsTarget = transitions.queryTarget.keys || transitions.queryTarget.elements
+        for (const key in transitionsTarget) {
+          let isKeyValid
+          if (transitions.queryTarget.keys) {
+            isKeyValid = !transitionsTarget[key].virtual
```

Comment:
It is good to see that this code was re written as it exposed the flaw of the original implementation.

It is not possible to select `virtual` and `association` fields as they don't exist on the database. So no matter whether it is the `keys` or the `elements` being processed they will have to be removed from the `keys`.

To prevent these issues from causing bugs a PR was created to unify this check, but it currently cannot be merged as it breaks the `UPSERT` behavior. So please take over the check as defined [here](https://github.com/cap-js/cds-dbs/pull/408/files#diff-d5bbedeb9d69ec4e9a2a2cb63e32f88b5098749d0e2505bd8fefc6b6d33cf3eaR959)

Reply:
Change this code according to your suggestion